File: src/Organon/Tfl/Document.hs
module Organon.Tfl.Document
  ( -- * Types
    Document (..),
    ProofBlock (..),
    Premise (..),
    RefModifier (..),
    SrcPos (..),
    Located (..),

    -- * Parsing
    parseDocument,
    fromConcreteH,
  )
where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Text (Text)
import qualified Data.Text as T
import Organon.Common.Types (Located (..), SrcPos (..))
import Organon.Tfl.Hole (SignedTermH (..), StatementH (..))
import Organon.Tfl.Parser
  ( Parser,
    isNameChar,
    lexeme,
    lineComment,
    sc,
    statementHP,
    symbol,
  )
import Organon.Tfl.Types
import Text.Megaparsec
import Text.Megaparsec.Char
import qualified Text.Megaparsec.Char.Lexer as L

-- | A modifier applied to a reference premise.
data RefModifier
  = -- | Simple conversion: swap terms (valid for E, I).
    RefConv
  | -- | Conversion per accidens: swap terms + weaken (A→I, E→O).
    RefPerAccidens
  | -- | Obversion: flip one sign, complement its term (always valid).
    RefObv
  | -- | Contraposition: flip all signs, complement all terms (valid for A, O).
    RefContra
  deriving stock (Eq, Ord, Show)

-- | A premise in a proof block: either a concrete statement, a reference
-- to a previously proved conclusion, or a statement with holes.
data Premise
  = -- | A concrete TFL statement.
    PremiseStmt Statement
  | -- | A reference to a previously proved conclusion.
    --   Fields: optional namespace, proof name, optional modifier.
    PremiseRef (Maybe Text) Text (Maybe RefModifier)
  | -- | A statement containing holes.
    PremiseHole StatementH
  deriving stock (Eq, Show)

-- | A proof block: a named inference with premises and a conclusion.
data ProofBlock = ProofBlock
  { proofName :: Located Text,
    proofPremises :: [Located Premise],
    proofConclusion :: Located StatementH
  }
  deriving stock (Eq, Show)

-- | A TFL document: an optional list of open directives, an optional
-- relational term lexicon, and proof blocks.
-- No tradition directive — TFL has a single logic.
data Document = Document
  { docOpens :: [Located Text],
    docRelLexicon :: RelLexicon,
    docProofs :: [Located ProofBlock]
  }
  deriving stock (Eq, Show)

-- ---------------------------------------------------------------------------
-- Parsing
-- ---------------------------------------------------------------------------

-- | Parse a @.tfl@ document.
parseDocument :: Text -> Either Text Document
parseDocument input =
  case parse (scn *> documentP <* eof) "" input of
    Left err -> Left (T.pack (errorBundlePretty err))
    Right d -> Right d

-- Whitespace including newlines (for between top-level blocks).
scn :: Parser ()
scn = L.space space1 lineComment empty

-- Position helpers

getPos :: Parser SrcPos
getPos = do
  sp <- getSourcePos
  pure (SrcPos (unPos (sourceLine sp)) (unPos (sourceColumn sp)))

located :: Parser a -> Parser (Located a)
located p = do
  s <- getPos
  v <- p
  e <- getPos
  pure (Located s e v)

-- ---------------------------------------------------------------------------
-- Premise parsers
-- ---------------------------------------------------------------------------

premiseP :: Parser Premise
premiseP = refP <|> stmtPremiseP

stmtPremiseP :: Parser Premise
stmtPremiseP = do
  sh <- statementHP
  case fromConcreteH sh of
    Just stmt -> pure (PremiseStmt stmt)
    Nothing -> pure (PremiseHole sh)

refP :: Parser Premise
refP = do
  _ <- char '@'
  first <- lexeme (takeWhile1P (Just "proof name") isNameChar)
  mDot <- optional (char '.')
  case mDot of
    Just _ -> do
      name <- lexeme (takeWhile1P (Just "proof name") isNameChar)
      modifier <- refModifierP
      pure (PremiseRef (Just first) name modifier)
    Nothing -> do
      modifier <- refModifierP
      pure (PremiseRef Nothing first modifier)

refModifierP :: Parser (Maybe RefModifier)
refModifierP =
  optional
    ( RefConv <$ symbol "conv"
        <|> RefPerAccidens <$ symbol "per-accidens"
        <|> RefObv <$ symbol "obv"
        <|> RefContra <$ symbol "contra"
    )

-- | Extract a concrete 'Statement' from a 'StatementH', if it has no holes.
fromConcreteH :: StatementH -> Maybe Statement
fromConcreteH (StmtH sths) = Statement <$> traverse extractST sths
  where
    extractST (ConcreteSTH st) = Just st
    extractST HoleSTH = Nothing
fromConcreteH WholeStmtH = Nothing

-- ---------------------------------------------------------------------------
-- Proof block parser
-- ---------------------------------------------------------------------------

proofBlockP :: Parser ProofBlock
proofBlockP = do
  _ <- symbol "proof"
  name <- located (lexeme (takeWhile1P (Just "proof name") isNameChar))
  _ <- eol *> scn
  prems <- many (try (located premiseP <* eol <* scn))
  concl <- conclusionP
  pure (ProofBlock name prems concl)

conclusionP :: Parser (Located StatementH)
conclusionP = do
  _ <- symbol "∴" <|> symbol "therefore"
  located statementHP

-- ---------------------------------------------------------------------------
-- Open directive
-- ---------------------------------------------------------------------------

openP :: Parser (Located Text)
openP = located $ do
  _ <- symbol "open"
  lexeme (takeWhile1P (Just "namespace name") isNameChar)

-- ---------------------------------------------------------------------------
-- Rel directive
-- ---------------------------------------------------------------------------

-- | Parse a @rel@ directive: @rel Love "Lover of" "Loved by"@.
relP :: Parser (Text, RelForms)
relP = do
  _ <- symbol "rel"
  name <- lexeme (takeWhile1P (Just "term name") isNameChar)
  active <- quotedStringP
  _ <- sc
  passive <- quotedStringP
  pure (name, RelForms active passive)

-- | Parse a double-quoted string literal.
quotedStringP :: Parser Text
quotedStringP = do
  _ <- char '"'
  content <- takeWhileP (Just "string character") (/= '"')
  _ <- char '"'
  pure content

-- ---------------------------------------------------------------------------
-- Document parser
-- ---------------------------------------------------------------------------

documentP :: Parser Document
documentP = do
  opens <- many (try (openP <* eol <* scn))
  rels <- many (try (relP <* eol <* scn))
  proofs <- many (located proofBlockP <* scn)
  pure (Document opens (Map.fromList rels) proofs)

File: web/src/core/tfl/document.ts
// TFL Document parser — ported from Organon.Tfl.Document

import type { StatementH, SignedTermH } from "./hole";
import {
  ParseError,
  ParserState,
  isNameChar,
  isTflTermChar,
  statementHP,
} from "./parser";
import type { Statement } from "./types";
import type { RelForms, RelLexicon } from "./types";

// ---------------------------------------------------------------------------
// Source positions & Located
// ---------------------------------------------------------------------------

export type SrcPos = {
  readonly posLine: number;
  readonly posCol: number;
};

export type Located<T> = {
  readonly locStart: SrcPos;
  readonly locEnd: SrcPos;
  readonly locValue: T;
};

// ---------------------------------------------------------------------------
// Premise types
// ---------------------------------------------------------------------------

export type RefModifier = "conv" | "per-accidens" | "obv" | "contra";

export type Premise =
  | { readonly tag: "PremiseStmt"; readonly stmt: Statement }
  | {
      readonly tag: "PremiseRef";
      readonly namespace: string | null;
      readonly name: string;
      readonly modifier: RefModifier | null;
    }
  | { readonly tag: "PremiseHole"; readonly stmtH: StatementH };

// ---------------------------------------------------------------------------
// Proof block & Document
// ---------------------------------------------------------------------------

export type ProofBlock = {
  readonly proofName: Located<string>;
  readonly proofPremises: Located<Premise>[];
  readonly proofConclusion: Located<StatementH>;
};

export type Document = {
  readonly docOpens: Located<string>[];
  readonly docRelLexicon: RelLexicon;
  readonly docProofs: Located<ProofBlock>[];
};

// ---------------------------------------------------------------------------
// fromConcreteH
// ---------------------------------------------------------------------------

export function fromConcreteH(sh: StatementH): Statement | null {
  if (sh.tag === "WholeStmtH") return null;
  const terms: import("./types").SignedTerm[] = [];
  for (const sth of sh.terms) {
    if (sth.tag === "HoleSTH") return null;
    terms.push(sth.st);
  }
  return { terms };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function getPos(p: ParserState): SrcPos {
  return { posLine: p.line, posCol: p.col };
}

function located<T>(p: ParserState, parse: () => T | null): Located<T> | null {
  const start = getPos(p);
  const value = parse();
  if (value === null) return null;
  const end = getPos(p);
  return { locStart: start, locEnd: end, locValue: value };
}

// ---------------------------------------------------------------------------
// Premise parsers
// ---------------------------------------------------------------------------

function refModifierP(p: ParserState): RefModifier | null {
  if (p.symbol("conv")) return "conv";
  if (p.symbol("per-accidens")) return "per-accidens";
  if (p.symbol("obv")) return "obv";
  if (p.symbol("contra")) return "contra";
  return null;
}

function refP(p: ParserState): Premise | null {
  const s = p.save();
  if (p.eof || p.peek() !== "@") {
    p.restore(s);
    return null;
  }
  p.advance(); // consume '@'
  const firstName = p.takeWhile1(isNameChar);
  if (firstName === null) {
    p.restore(s);
    return null;
  }
  p.sc();

  // Check for qualified name (namespace.name)
  if (!p.eof && p.peek() === ".") {
    p.advance();
    const secondName = p.takeWhile1(isNameChar);
    if (secondName === null) {
      p.restore(s);
      return null;
    }
    p.sc();
    const modifier = refModifierP(p);
    return {
      tag: "PremiseRef",
      namespace: firstName,
      name: secondName,
      modifier,
    };
  }

  const modifier = refModifierP(p);
  return { tag: "PremiseRef", namespace: null, name: firstName, modifier };
}

function stmtPremiseP(p: ParserState): Premise | null {
  const sh = statementHP(p);
  if (sh === null) return null;
  const concrete = fromConcreteH(sh);
  if (concrete !== null) return { tag: "PremiseStmt", stmt: concrete };
  return { tag: "PremiseHole", stmtH: sh };
}

function premiseP(p: ParserState): Premise | null {
  const s = p.save();
  const ref = refP(p);
  if (ref !== null) return ref;
  p.restore(s);
  return stmtPremiseP(p);
}

// ---------------------------------------------------------------------------
// Proof block parser
// ---------------------------------------------------------------------------

function proofBlockP(p: ParserState): ProofBlock | null {
  if (!p.symbol("proof")) return null;
  const nameStart = getPos(p);
  const name = p.takeWhile1(isNameChar);
  if (name === null) return null;
  p.sc();
  const nameEnd = getPos(p);
  const proofName: Located<string> = {
    locStart: nameStart,
    locEnd: nameEnd,
    locValue: name,
  };

  // Expect newline after proof name
  if (!p.newline()) return null;
  p.scn();

  // Parse premises (lines before ∴)
  const premises: Located<Premise>[] = [];
  while (true) {
    const s = p.save();
    // Check if we've hit the conclusion marker
    if (p.symbol("∴") || p.symbol("therefore")) {
      p.restore(s);
      break;
    }
    p.restore(s);
    if (p.eof) break;

    const lp = located(p, () => premiseP(p));
    if (lp === null) break;
    premises.push(lp);
    p.newline();
    p.scn();
  }

  // Parse conclusion: ∴ <statementH>
  if (!p.symbol("∴") && !p.symbol("therefore")) return null;
  const conclStart = getPos(p);
  const conclSH = statementHP(p);
  if (conclSH === null) return null;
  const conclEnd = getPos(p);
  const proofConclusion: Located<StatementH> = {
    locStart: conclStart,
    locEnd: conclEnd,
    locValue: conclSH,
  };

  return { proofName, proofPremises: premises, proofConclusion };
}

// ---------------------------------------------------------------------------
// Open directive
// ---------------------------------------------------------------------------

function openP(p: ParserState): string | null {
  if (!p.symbol("open")) return null;
  const name = p.takeWhile1(isNameChar);
  if (name === null) return null;
  p.sc();
  return name;
}

// ---------------------------------------------------------------------------
// Rel directive parser
// ---------------------------------------------------------------------------

function quotedStringP(p: ParserState): string | null {
  if (p.eof || p.peek() !== '"') return null;
  p.advance();
  let result = "";
  while (!p.eof && p.peek() !== '"') {
    result += p.peek();
    p.advance();
  }
  if (p.eof) return null;
  p.advance(); // closing quote
  p.sc();
  return result;
}

function relP(p: ParserState): [string, RelForms] | null {
  const s = p.save();
  if (!p.symbol("rel")) {
    p.restore(s);
    return null;
  }
  const name = p.takeWhile1(isTflTermChar);
  if (name === null) {
    p.restore(s);
    return null;
  }
  p.sc();
  const active = quotedStringP(p);
  if (active === null) {
    p.restore(s);
    return null;
  }
  const passive = quotedStringP(p);
  if (passive === null) {
    p.restore(s);
    return null;
  }
  return [name, { relActive: active, relPassive: passive }];
}

// ---------------------------------------------------------------------------
// Document parser
// ---------------------------------------------------------------------------

function documentP(p: ParserState): Document {
  const opens: Located<string>[] = [];
  const rels: [string, RelForms][] = [];
  const proofs: Located<ProofBlock>[] = [];

  p.scn();

  // Parse open directives
  while (true) {
    const s = p.save();
    const lo = located(p, () => openP(p));
    if (lo === null) {
      p.restore(s);
      break;
    }
    opens.push(lo);
    p.newline();
    p.scn();
  }

  // Parse rel directives
  while (true) {
    const s = p.save();
    const r = relP(p);
    if (r === null) {
      p.restore(s);
      break;
    }
    rels.push(r);
    p.newline();
    p.scn();
  }

  // Parse proof blocks
  while (!p.eof) {
    const lp = located(p, () => proofBlockP(p));
    if (lp === null) break;
    proofs.push(lp);
    p.scn();
  }

  return {
    docOpens: opens,
    docRelLexicon: new Map(rels),
    docProofs: proofs,
  };
}

export function parseDocument(input: string): Document | ParseError {
  const p = new ParserState(input);
  const doc = documentP(p);
  if (!p.eof) {
    return new ParseError(
      p.line,
      p.col,
      `Unexpected input at line ${p.line}, column ${p.col}`,
    );
  }
  return doc;
}
